### DISCUSSION TITLE
Add Plugins

### DISCUSSION THREAD

**Discussion by rocky4546** [cabernetwork/cabernet#112] [Ideas]
Discussions on adding different plugins.

**Comment by rocky4546** [comment]
Wondering if it's possible for cabernet to work with https://thetvapp.to/tv? Would this require a new plug-in similar to the existing ones? The daddylive site seems to have lots of issues / go down a lot and tvapp website is more reliable for me and would love to be able to pipe it into plex using cabernet if possible

**Reply by slippyC** [reply]
There's a Kodi plugin for that site(currently).  So at least it wouldn't be to hard to figure out how to get the actual streams.  The main advantage DaddyLive has is more channels.  I would prefer reliability over more, personally.  There is really only 1 channel on DaddyLive that I would want that isn't available on TheTVApp site.

**Comment by cookieisland** [comment]
The channels on thetvapp seem to be the same as what we have already in cabernet with Pluto and Xumo. 

The nice thing about DaddyLive is access to channels from various other countries. E.g. the UK with BBC One, Two, Three, Four, ITV1-4, etc. Also it has Canadian channels like CBC, CTV,etc.  These are definitely channels I watch and record. While they do not have a tvguide/epg for these UK and Canadian channels provided by cabernet you can pull the EPG from at least two or three different places that I know if and the EPG is free as long as you do a bit of scripting and cron jobs (mc2xml, webgrabplus and/or zap2xml)

**Reply by slippyC** [reply]
I don't think you'll find any of the channels on TheTVApp on Pluto or Xumo.  Those are "Premium" US channels.  Ever heard of HBO, Cinemax, ESPN, etc?  If you can get those through Pluto or Xumo, please point me in the direction to that app(which none of those really matter that much to me, but it's an example). 

One other thing.  I believe TheTVApp even has some channels that DaddyLive doesn't(as far as Premium).  I agree that DaddyLive has a lot more, but it can be very sporadic at times like originally posted.

**Reply by Aetherinox** [reply]
I have both DaddyLive and TheTvApp in my list, simply to have redundancy. You never know when one service will die.

However, what I did is I created a docker image which automatically does the fetching to get the latest EPG and m3u for thetvapp, and then I add a new m3u into Cabernet. It has worked great.

https://github.com/Aetherinox/thetvapp-docker

Yeah it's not a direct Cabernet plugin, but I'm running it on cabernet and haven't had an issue.

```ini
[m3u_default]
label = TheTvApp Instance
enabled = True
channel-m3u_file = https://tv.domain.com/thetvapp.m3u8
channel-start_ch_num = 700
player-stream_type = m3u8redirect
epg-xmltv_file = https://tv.domain.com/thetvapp.xml.gz
```

**Reply by ccihon** [reply]
Sorry to resurrect this older discussion but I got the TVApp-2 docker up and working great - is there a way to have it pull an epg longer than the 48 hours or so that it does?  Thanks

**Comment by cookieisland** [comment]
I have watched and recorded HBO, Cinemax, etc via Cabernet so I should have included DaddyLive with Pluto and Xumo in my first statement.  DaddyLive has HBO, Cinemax, and more.  The problem with DaddyLive is there has been issues (timeout, outages,etc).  However I do watch and record multiple DaddyLive channels these days and it is not too bad with some tweaks I am testing on my tvheadend configuration.  I am hoping DaddyLive improves and/or we find ways to mitigate the issues.

**Comment by bkwsoft** [comment]
Is there any kind of guide or documentation available on plug-in development?

**Comment by rocky4546** [comment]
Sorry, no documentation. I normally just take a plugin and mod it.  Best one to use is probably the xumo plugin.

**Reply by bkwsoft** [reply]
So I spent the weekend to stand up my first attempt at a plugin and ran into a potential bug so I'm wondering if anyone has successfully implemented a plugin outside of the cabernet managed repository.

When putting my plugin into the plugins_ext folder, it loads, but then subsequently throws a SQL Constraint violation on the plugins table for a null repo value.   It comes down to line line 188 of lib/plugins/plugin.py 

`json_settings['repoid'] = None`

Essentially the repoID is always set to None even if I were to specify a repoId in my plugin.json file.

I've tried two different "fixes" one being to change the repo column on the plugin and plugin_instance tables.   That gets the plugin to fully load, but playback doesn't like to work http://192.168.1.239:5004/{namespace}/watch/161 immediately fails saying the channel doesn't exist yet http://192.168.1.239:5004/watch/161 works just fine.

Second I updated the plugin.py file replacing the `json_settings['repoid'] = None` with `self.repo_id = json_settings['repoid']`.  This lets the plugin load again, and playback including the namespace in the path works as expected.

**Reply by markcs** [reply]
Did you ever solve this issue?

**Reply by bkwsoft** [reply]
I have not to date.   I’ve been running with a patched version for a while now and haven’t put much more effort into it to make it more widely usable yet.

I’m basically using TV Everywhere authentication to put live TV streams, but it’s currently specific to my provider where it needs to be configurable to any provider before it really makes sense for more general availability.

**Comment by rocky4546** [comment]
have you pushed the plugin to a repo somewhere?

**Reply by bkwsoft** [reply]
I have, it’s still very much a work in progress and not ready for general consumption.

**Reply by bkwsoft** [reply]
FYI, I did send you and invite to the repo.

**Comment by rocky4546** [comment]
Now that DL is up and running to the most extent, I plan on working this area, but will be overwhelmed for the next week.

**Comment by rocky4546** [comment]
https://github.com/lem85930/thetvapp-m3u/tree/main

**Comment by rocky4546** [comment]
Fixed cabernet so you can add plugins without repos and have them run and test out.  External repos are still not available, but stand alone plugins are working fine.
